text/0006-sampling.md

```diff
@@ -0,0 +1,302 @@
+![Personas](https://i.imgur.com/w1H0CfH.png)
+
+## Explanation
+
+We outline five different use cases (who may be overlapping sets of people), and how they should
+interact with OpenTelemetry:
+
+### Library developer
+Examples: gRPC, Express, Django developers.
+
+ * They must only depend upon the OpenTelemetry API and not upon the SDK.
+   * For testing only they may depend on the SDK with InMemoryExporter.
+ * They are shipping source code that will be linked into others' applications.
+ * They have no explicit runtime control over the application.
+ * They know some signal about what traces may be interesting (e.g. unusual control plane requests)
+ or uninteresting (e.g. health-checks), but have to write fully generically.
+
+**Solution:**
+
+ * On the start Span operation, the OpenTelemetry API will allow marking a span with one of three
+ choices for the [SamplingHint](#samplinghint).
+ 
+### Infrastructure package/binary developer
+Examples: HBase, Envoy developers.
+
+ * They are shipping self-contained binaries that may accept YAML or similar run-time configuration,
+ but are not expected to support extensibility/plugins beyond the default OpenTelemetry SDK, 
+ OpenTelemetry SDKTracer, and OpenTelemetry wire format exporter.
+ * They may have their own recommendations for sampling rates, but don't run the binaries in
+ production, only provide packaged binaries. So their sampling rate configs, and sampling strategies
+ need to a finite "built in" set from OpenTelemetry's SDK.
+ * They need to deal with upstream sampling decisions made by services calling them.
+
+**Solution:**
+ * Allow different sampling strategies by default in OpenTelemetry SDK, all configurable easily via
+ YAML or future flags. See [default samplers](#default-samplers).
+
+### Application developer
+These are the folks we've been thinking the most about for OpenTelemetry in general.
+
+ * They have full control over the OpenTelemetry implementation or SDK configuration. When using the
+ SDK they can configure custom exporters, custom code/samplers, etc.
+ * They can choose to implement runtime configuration via a variety of means (e.g. baking in feature
+ flags, reading YAML files, etc.), or even configure the library in code.
+ * They make heavy usage of OpenTelemetry for instrumenting application-specific behavior, beyond
+ what may be provided by the libraries they use such as gRPC, Django, etc.
+
+**Solution:**
+ * Allow application developers to link in custom samplers or write their own when using the
+ official SDK.
+   * These might include dynamic per-field sampling to achieve a target rate
+   (e.g. https://github.com/honeycombio/dynsampler-go)
+ * Sampling decisions are made within the start Span operation, after attributes relevant to the
+ span have been added to the Span start operation but before a concrete Span object exists (so that
+ either a NoOpSpan can be made, or an actual Span instance can be produced depending upon the
+ sampler's decision).
```

Comment:
ad-hoc sampling rules seem like they're ultimately a property of the collection exporter/backend rather than of the OTel SDK... e.g. I wouldn't insist on making the OTel SDK exactly replicate what Honeycomb does post-collection...

Reply:
@yurishkuro can you clarify if the sampling decision still happens on the client side? If that is true can you clarify when is the limit when that needs to happen. One possibility that I covered in the "when does sampling happen" is that we can delay until the moment the SpanContext is accessed: to create a child, to serialize it, or for example to do log correlation.

Reply:
Yes, sampling decision still happens in the client.

The limit depends on the type of sampling policies provided to the client. Consider these policies:
  * A: static: sample with probability p=0.1
  * B: if span.name = "/userProfile" then sample with rate limit r=10/sec
  * C: tag based: if tag X = Y, then sample with probability p=0.9

If the tracer is configured with policies [A], then sampling decision is finalized immediately on span creation, due to the nature of the sole policy A which is independent of any later span data.

If the tracer is configured with policies [B, A], then sampling decision can happen on span creation if A says Yes or B says Yes, or after span.setName() (assumption is that the code is only going to call setName once, at least as far as sampling decision is concerned).

If tracer is configured with all 3 policies, then recording of events may never stop if none of the policies match and/or say Yes. It's more expensive but it's the user's choice.

> we can delay until the moment the SpanContext is accessed

That was how we initially implemented the handling of late span.SetName() binding. However, there's no particular reason why the sampling on the parent span cannot remain "undecided" even if child spans are being created. Yes, it will result in some spans of the trace not captured, but that's just the nature of the combination of the chosen sampling policies. We're also thinking that we will use a shared sampling state across all spans for the trace in memory, so  that some of those sampling policies might fire on a child span and trigger sampling of the parent.

Disclaimer: we just brainstormed these scenarios and will prototype them in the next week or so. There's a chance it might not work at all, but so far it looks this approach is meeting our requirements (which definitely include late sampling decision based on a specific tag).

Reply:
With respect, "we will prototype 'undecided' states in the next week" feels like "add to v0.0.2 but omit for v0.0.1" given Sep 1 spec freeze. Believe me, I'm all for late-binding sampling decisions, but we need to be explicit about when there is the possibility of that happening, and allowing implementations to act accordingly. We can always add complexity in a methodical fashion, I'm worried about rushing and getting it wrong.

Reply:
The solution with late binding of span name alone has been in Jaeger in production for >2yrs, implemented as Bogdan implied - events are recorded until sampling decision is finalized, which happens either on span.SetName() or a child span being created. I don't particularly like it, but we had some Node.js frameworks where span name was only known later. The approach I described above is meant to generalize and improve it.

I don't have a problem with not supporting any form of late binding in v1, but this RFC does provide for it via "late Builder", which feels even worse to me.

Reply:
Ah, I think I understand. Hm. I often know the attributes "at span creation", but I want the time to compute the attributes to be included in the span duration, if that makes sense? That's why we settled on the builder pattern as it was easier for application programmers to work with...

Is there some middle ground here that allows very-short-delayed binding without requiring it all upfront at span creation and adjusting the times afterward?

Reply:
I guess the question is how big of a deal it would be if v1 didn't support any form of late binding and punted it on v2. In Jaeger, only Node.js client supports late span name binding that impacts sampling, nobody complained in other languages (it only matters when using sampling policies sensitive to span name).

Reply:
I am fine to remove the delayed sampling for the moment. But I will still use as an argument in some of the decisions that we may support this, for example the reason we have a different isRecordingEvents is because we may implement something like isDelayed that you suggested so we don't want people to play with the trace flags to know if they should record or not something on the Span.